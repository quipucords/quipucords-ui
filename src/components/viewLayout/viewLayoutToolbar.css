.quipucords-toolbar__user-dropdown {
  align-items: center;
  display: flex;
  gap: var(--pf-v5-global--spacer--sm);
}
.quipucords-toolbar__user-dropdown .pf-v5-c-avatar {
  width: 36px;
  height: 36px;
  background: url(../../images/imgAvatar.svg);
}
