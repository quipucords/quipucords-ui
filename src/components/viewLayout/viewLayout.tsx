/**
 * Provides the main layout for the application using PatternFly, incorporating a header, navigation sidebar,
 * and content area. It dynamically adjusts navigation based on routing, supports a sidebar toggle, and
 * ensures accessibility with a "Skip to Content" link. Excludes header and sidebar on the login page for a
 * simplified UI.
 *
 * @module appLayout
 */
import * as React from 'react';
import { useTranslation } from 'react-i18next';
import { NavLink, useLocation } from 'react-router-dom';
import {
  Brand,
  Button,
  Masthead,
  MastheadBrand,
  MastheadContent,
  MastheadMain,
  MastheadToggle,
  Nav,
  NavExpandable,
  NavItem,
  NavList,
  Page,
  PageSidebar,
  PageSidebarBody,
  SkipToContent
} from '@patternfly/react-core';
import { BarsIcon } from '@patternfly/react-icons';
import { helpers } from '../../helpers';
import { IAppRoute, IAppRouteGroup, routes } from '../../routes';
import { AppToolbar } from './viewLayoutToolbar';

interface AppLayoutProps {
  children: React.ReactNode;
  titleImg?: string;
  uiName?: string;
}

const AppLayout: React.FC<AppLayoutProps> = ({
  children,
  titleImg = helpers.getTitleImg(),
  uiName = helpers.UI_NAME
}) => {
  const { t } = useTranslation();
  const [sidebarOpen, setSidebarOpen] = React.useState(true);

  const Header = (
    <Masthead>
      <MastheadToggle>
        <Button variant="plain" onClick={() => setSidebarOpen(!sidebarOpen)} aria-label="Global navigation">
          <BarsIcon />
        </Button>
      </MastheadToggle>
      <MastheadMain>
        <MastheadBrand>
          <Brand alt={t('view.alt-logo', { name: uiName })} heights={{ default: '36px' }}>
            <source srcSet={titleImg} />
          </Brand>
        </MastheadBrand>
      </MastheadMain>
      <MastheadContent>
        <AppToolbar />
      </MastheadContent>
    </Masthead>
  );

  const location = useLocation();

  const renderNavItem = (route: IAppRoute, index: number) => (
    <NavItem key={`${route.label}-${index}`} id={`${route.label}-${index}`} isActive={route.path === location.pathname}>
      <NavLink to={route.path}>{route.label}</NavLink>
    </NavItem>
  );

  const renderNavGroup = (group: IAppRouteGroup, groupIndex: number) => (
    <NavExpandable
      key={`${group.label}-${groupIndex}`}
      id={`${group.label}-${groupIndex}`}
      title={group.label}
      isActive={group.routes.some(route => route.path === location.pathname)}
    >
      {group.routes.map((route, idx) => route.label && renderNavItem(route, idx))}
    </NavExpandable>
  );

  // FixMe: PF spelling bug in attr "forwardScrollAriaLabel"
  const Navigation = (
    <Nav id="nav-primary-simple" theme="dark">
      <NavList id="nav-list-simple" forwardScrollAriaLabel="Scroll forward">
        {routes.map(
          (route, idx) => route.label && (!route.routes ? renderNavItem(route, idx) : renderNavGroup(route, idx))
        )}
      </NavList>
    </Nav>
  );

  const Sidebar = (
    <PageSidebar theme="dark">
      <PageSidebarBody>{Navigation}</PageSidebarBody>
    </PageSidebar>
  );

  const pageId = 'primary-app-container';

  const PageSkipToContent = (
    <SkipToContent
      onClick={event => {
        event.preventDefault();
        const primaryContentContainer = document.getElementById(pageId);
        if (primaryContentContainer) {
          primaryContentContainer.focus();
        }
      }}
      href={`#${pageId}`}
    >
      Skip to Content
    </SkipToContent>
  );

  return (
    <Page mainContainerId={pageId} header={Header} sidebar={sidebarOpen && Sidebar} skipToContent={PageSkipToContent}>
      {children}
    </Page>
  );
};

export { AppLayout as default, AppLayout, type AppLayoutProps };
