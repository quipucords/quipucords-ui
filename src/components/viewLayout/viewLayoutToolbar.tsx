/**
 * Utilizes PatternFly for a responsive toolbar with theme toggling, help, and user logout functionality.
 * It auto-detects theme preference, provides a dropdown for help and user actions, and manages session state for logout.
 * @module appToolbar
 */
import * as React from 'react';
import { useNavigate } from 'react-router-dom';
import {
  Dropdown,
  DropdownItem,
  Icon,
  MenuToggle,
  ToggleGroup,
  ToggleGroupItem,
  Toolbar,
  ToolbarContent,
  ToolbarGroup,
  ToolbarItem
} from '@patternfly/react-core';
import { EllipsisVIcon, MoonIcon, QuestionCircleIcon, SunIcon } from '@patternfly/react-icons';
import axios from 'axios';
import { useUsername } from '../sessionContext/sessionProvider';
import '@patternfly/react-styles/css/components/Avatar/avatar.css';
import './viewLayoutToolbar.css';

const AppToolbar: React.FunctionComponent = () => {
  const [helpOpen, setHelpOpen] = React.useState<boolean>(false);
  const [userDropdownOpen, setUserDropdownOpen] = React.useState<boolean>(false);
  const [kebabDropdownOpen, setKebabDropdownOpen] = React.useState<boolean>(false);
  const [isDarkTheme, setIsDarkTheme] = React.useState(
    window.matchMedia && window.matchMedia('(prefers-color-scheme: dark)').matches
  );

  const applyTheme = isDark => {
    const htmlElement = document.getElementsByTagName('html')[0];
    if (htmlElement) {
      if (isDark) {
        htmlElement.classList.add('pf-v5-theme-dark');
      } else {
        htmlElement.classList.remove('pf-v5-theme-dark');
      }
    }
  };
  applyTheme(isDarkTheme);

  const userName = useUsername();
  const nav = useNavigate();

  const onAbout = () => {};

  const onLogout = () => {
    axios
      .put(`${process.env.REACT_APP_USER_SERVICE_LOGOUT}`)
      .catch(err => {
        console.error('Failed to logout', err);
      })
      .finally(() => {
        localStorage.removeItem('authToken');
        nav('/login');
      });
  };

  const onHelpSelect = (
    _event: React.MouseEvent<Element, MouseEvent> | undefined,
    value: string | number | undefined
  ) => {
    // eslint-disable-next-line no-console
    console.log('selected', value);
    setHelpOpen(false);
  };

  const onUserDropdownSelect = (
    _event: React.MouseEvent<Element, MouseEvent> | undefined,
    value: string | number | undefined
  ) => {
    // eslint-disable-next-line no-console
    console.log('selected', value);
    setUserDropdownOpen(false);
  };

  return (
    <Toolbar id="toolbar" isFullHeight isStatic>
      <ToolbarContent>
        <ToolbarGroup
          variant="icon-button-group"
          align={{ default: 'alignRight' }}
          spacer={{ default: 'spacerNone', md: 'spacerMd' }}
        >
          <ToolbarGroup
            variant="icon-button-group"
            visibility={{ default: 'hidden', lg: 'visible' }}
          >
            <ToolbarItem>
              <ToggleGroup aria-label="Dark theme toggle group">
                <ToggleGroupItem
                  aria-label="light theme toggle"
                  icon={
                    <Icon size="md">
                      <SunIcon />
                    </Icon>
                  }
                  isSelected={!isDarkTheme}
                  onChange={() => {
                    setIsDarkTheme(false);
                    applyTheme(false);
                  }}
                />
                <ToggleGroupItem
                  aria-label="dark theme toggle"
                  icon={
                    <Icon size="md">
                      <MoonIcon />
                    </Icon>
                  }
                  isSelected={isDarkTheme}
                  onChange={() => {
                    setIsDarkTheme(true);
                    applyTheme(true);
                  }}
                />
              </ToggleGroup>
            </ToolbarItem>
            <ToolbarItem>
              <Dropdown
                popperProps={{ position: 'right' }}
                onSelect={onHelpSelect}
                onOpenChange={(isOpen: boolean) => setHelpOpen(isOpen)}
                isOpen={helpOpen}
                toggle={toggleRef => (
                  <MenuToggle
                    aria-label="Toggle"
                    ref={toggleRef}
                    variant="plain"
                    onClick={() => setHelpOpen(prev => !prev)}
                    isExpanded={helpOpen}
                  >
                    <QuestionCircleIcon />
                  </MenuToggle>
                )}
              >
                <DropdownItem onClick={onAbout} value="about">
                  About
                </DropdownItem>
              </Dropdown>
            </ToolbarItem>
          </ToolbarGroup>
          <ToolbarItem visibility={{ default: 'visible', lg: 'hidden' }}>
            <Dropdown
              isPlain
              popperProps={{ position: 'right' }}
              onSelect={onUserDropdownSelect}
              onOpenChange={(isOpen: boolean) => setKebabDropdownOpen(isOpen)}
              isOpen={kebabDropdownOpen}
              toggle={toggleRef => (
                <MenuToggle
                  aria-label="Toggle"
                  ref={toggleRef}
                  variant="plain"
                  onClick={() => setKebabDropdownOpen(prev => !prev)}
                  isExpanded={kebabDropdownOpen}
                  style={{ width: 'auto' }}
                >
                  <EllipsisVIcon />
                </MenuToggle>
              )}
              ouiaId="user_dropdown"
            >
              <DropdownItem value="logout" onClick={onLogout}>
                Logout
              </DropdownItem>
            </Dropdown>
          </ToolbarItem>
        </ToolbarGroup>
        <ToolbarItem visibility={{ default: 'hidden', lg: 'visible' }}>
          <Dropdown
            onSelect={onUserDropdownSelect}
            onOpenChange={(isOpen: boolean) => setUserDropdownOpen(isOpen)}
            isOpen={userDropdownOpen}
            ouiaId="user_dropdown"
            toggle={toggleRef => (
              <MenuToggle
                aria-label="Toggle"
                ref={toggleRef}
                variant="plain"
                onClick={() => setUserDropdownOpen(prev => !prev)}
                isExpanded={userDropdownOpen}
              >
                <div className="quipucords-toolbar__user-dropdown">
                  <span className="pf-v5-c-avatar" />
                  {userName}
                </div>
              </MenuToggle>
            )}
          >
            <DropdownItem value="logout" onClick={onLogout}>
              Logout
            </DropdownItem>
          </Dropdown>
        </ToolbarItem>
      </ToolbarContent>
    </Toolbar>
  );
};

export default AppToolbar;
