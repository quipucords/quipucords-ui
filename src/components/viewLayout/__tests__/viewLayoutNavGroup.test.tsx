import React from 'react';
import { shallowComponent } from '../../../../config/jest.setupTests';
jest.mock('../../../routes', () => {
  const actual = jest.requireActual('../../../routes');
  return {
    ...actual,
    routes: [
      {
        id: 'credentials',
        component: <div></div>,
        label: 'Credentials',
        path: '/credentials',
        title: 'Credentials'
      },
      {
        id: 'scans',
        label: 'Scans',
        routes: [
          {
            id: 'scans-results',
            component: <div></div>,
            label: 'Results',
            path: '/scans-results',
            title: 'Scans results'
          },
          {
            id: 'scans-running',
            component: <div></div>,
            label: 'Running',
            path: '/scans-running',
            title: 'Scans running'
          }
        ]
      }
    ]
  };
});
import { AppLayout as ViewLayout } from '../viewLayout';

describe('ViewLayout', () => {
  it('should render a basic component with NavGroup', async () => {
    const props = {
      children: 'Lorem ipsum'
    };
    const component = await shallowComponent(<ViewLayout {...props} />);
    expect(component).toMatchSnapshot('basic');
  });
});
