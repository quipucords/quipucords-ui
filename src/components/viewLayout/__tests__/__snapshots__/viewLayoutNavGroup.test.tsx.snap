// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewLayout should render a basic component with NavGroup: basic 1`] = `
<Page
  defaultManagedSidebarIsOpen={true}
  getBreakpoint={[Function]}
  getVerticalBreakpoint={[Function]}
  header={
    <Masthead>
      <MastheadToggle>
        <Button
          aria-label="Global navigation"
          onClick={[Function]}
          ouiaId="global-navigation"
          variant="plain"
        >
          <BarsIcon />
        </Button>
      </MastheadToggle>
      <MastheadMain>
        <MastheadBrand>
          <Brand
            alt="t([
  "view.alt-logo",
  {
    "name": "Quipucords"
  }
])"
            heights={
              {
                "default": "36px",
              }
            }
          >
            <source
              srcSet="title.svg"
            />
          </Brand>
        </MastheadBrand>
      </MastheadMain>
      <MastheadContent>
        <AppToolbar />
      </MastheadContent>
    </Masthead>
  }
  isBreadcrumbWidthLimited={false}
  isManagedSidebar={false}
  isNotificationDrawerExpanded={false}
  mainComponent="main"
  mainContainerId="primary-app-container"
  mainTabIndex={-1}
  onNotificationDrawerExpand={[Function]}
  sidebar={
    <PageSidebar
      theme="dark"
    >
      <PageSidebarBody>
        <Nav
          id="nav-primary-simple"
          onSelect={[Function]}
          onToggle={[Function]}
          ouiaSafe={true}
          theme="dark"
        >
          <NavList
            ariaLeftScroll="Scroll left"
            ariaRightScroll="Scroll right"
            backScrollAriaLabel="Scroll back"
            forwardScrollAriaLabel="Scroll forward"
            id="nav-list-simple"
          >
            <NavItem
              id="Credentials-0"
              isActive={false}
            >
              <NavLink
                to="/credentials"
              >
                Credentials
              </NavLink>
            </NavItem>
            <NavItem
              id="Scans-1"
              isActive={false}
            >
              <NavLink>
                Scans
              </NavLink>
            </NavItem>
          </NavList>
        </Nav>
      </PageSidebarBody>
    </PageSidebar>
  }
  skipToContent={
    <SkipToContent
      href="#primary-app-container"
      onClick={[Function]}
    >
      Skip to Content
    </SkipToContent>
  }
>
  Lorem ipsum
</Page>
`;
